mm2src/mm2_event_stream/src/streamer.rs

```diff
@@ -0,0 +1,230 @@
+use std::any::{self, Any};
+
+use crate::{Event, StreamingManager};
+use common::executor::{abortable_queue::WeakSpawner, AbortSettings, SpawnAbortable};
+use common::log::{error, info};
+
+use async_trait::async_trait;
+use futures::channel::{mpsc, oneshot};
+use futures::{future, select, FutureExt, Stream, StreamExt};
+
+/// A marker to indicate that the event streamer doesn't take any input data.
+pub struct NoDataIn;
+
+/// A mixture trait combining `Stream`, `Send` & `Unpin` together (to avoid confusing annotation).
+pub trait StreamHandlerInput<D>: Stream<Item = D> + Send + Unpin {}
+/// Implement the trait for all types `T` that implement `Stream<Item = D> + Send + Unpin` for any `D`.
+impl<T, D> StreamHandlerInput<D> for T where T: Stream<Item = D> + Send + Unpin {}
+
+#[async_trait]
+pub trait EventStreamer
+where
+    Self: Sized + Send + 'static,
+{
+    type DataInType: Send;
+
+    /// Returns a human readable unique identifier for the event streamer.
+    /// No other event streamer should have the same identifier.
+    fn streamer_id(&self) -> String;
+
+    /// Event handler that is responsible for broadcasting event data to the streaming channels.
+    ///
+    /// `ready_tx` is a oneshot sender that is used to send the initialization status of the event.
+    /// `data_rx` is a receiver that the streamer *could* use to receive data from the outside world.
+    async fn handle(
+        self,
+        broadcaster: Broadcaster,
+        ready_tx: oneshot::Sender<Result<(), String>>,
+        data_rx: impl StreamHandlerInput<Self::DataInType>,
+    );
+}
+
+/// Spawns the `EventStreamer::handle` in a separate thread.
```

Comment:
it would not be a thread for wasm. I guess spawns handle a future maybe is a more general phrase

Reply:
indeed, this actually should be named `task` instead.

Reply:
[52af46f](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2172/commits/52af46fe40523e897eaa0349c8398018cfb3435d)